File: code/1645.cpp
#include <bits/stdc++.h>

using namespace std;

int ans = 0;

long long consec_seq_util(int arr[], int n, int k, vector<vector<long long>> &dp)
{
	if(k == 1) {
		return 1;
	}
	long long ans = 0;
	for(int j = 1; j < n; j++) {
		if(arr[0] < arr[j]) {		
			if(dp[n-j][k-1] == -1) {
				dp[n-j][k-1] = consec_seq_util(&arr[j], n-j, k-1, dp);
			}
			ans += dp[n-j][k-1];
		}
	}
	return ans;
}

long long consec_seq(int arr[], int n, int k) 
{
	vector<vector<long long>> dp(n+1, vector<long long>(k+1,-1));
	long long ans = 0;
	for(int i = 0; i < n-k+1; i++) {
		ans += consec_seq_util(&arr[i], n-i, k, dp);
	}
	return ans;
}

int main()
{
	ios::sync_with_stdio(false);
	
	int N, K;

	while(cin >> N >> K && N != 0)
	{
		int arr[N];
	
		for(int i = 0; i < N; i++) {
			cin >> arr[i];
		}		
		
		cout << consec_seq(arr, N, K) << endl;
	}
}

File: code/Paradigmas/p-level-3/1645.py
/*
Nome:      El Dorado
ID:        1645
Resposta:  Accepted
Linguagem: Python 3 (Python 3.4.3) [+1s]
Tempo:     0.688s
Tamanho:   735 Bytes
Submissao: 05/06/16 20:03:38
*/
# -*- coding: utf-8 -*-
import functools

def cache(func):
	cache.cache = dict()
	def cached(*args):
		if args not in cache.cache:
			result = func(*args)
			cache.cache[args] = result
			return cache.cache[args]
		else:
			return cache.cache[args]
	return cached

@cache
def consec_seq_util(arr, n, k):
	if k == 1:
		return 1
	ans = 0
	for i in range(n):
		if arr[0] < arr[i]:
			ans += consec_seq_util(arr[i:], n-i, k-1)
	return ans

def consec_seq(arr, n, k):
	ans = 0
	for i in range(n-k+1):
		ans += consec_seq_util(arr[i:], n-i, k)
	return ans

while True:
	N, K = (int(i) for i in input().split())
	if N == 0:
		break
	arr = tuple(int(i) for i in input().split())
	print(consec_seq(arr, N, K))